Expose audio mixers to mods

hlysine opened the pull request:
Re: https://github.com/ppy/osu/discussions/16740#discussioncomment-2107632

The new `IApplicableToTrackMixer` interface mainly provides a way for mods to apply audio filters in song select. During gameplay, this interface is a shortcut to the already possible way of adding a drawable via `IApplicableToDrawableRuleset`, then getting the `AudioManager.TrackMixer` via DI.

This PR is a pre-requisite to changing muted mod to use a low-pass filter only, instead of adjusting track volume. 

This is a relatively simple diff, but there are some points I'm not super sure about:
- **Should `IApplicableToTrack` provide the mixer instead of a separate interface?** 
@frenzibyte originally proposed to include the mixer in `IApplicableToTrack.ApplyToTrack`, but there are many cases where a mixer isn't available (e.g. when using this interface to calculate new track rate in diff calc), so I decided to split this into a new interface.

- **Should `IApplicableToSampleMixer` exist?**
Yes for symmetry, but muted mod probably won't be needing this since it requires separate control of hit samples and metronome samples. Feel free to revert 016e0c320745dae61a9c273d4bbb46d382392baa if this is deemed redundant.

- **Should local mixers be provided to `ApplyTo*Mixer` instead?**
I don't think any local mixers exist yet. Just providing the global mixers should be the simplest working implementation IMO.


For reference, this is how the muted mod could look like using this interface:
<details>
<summary>diff</summary>

```diff
Index: osu.Game/Rulesets/Mods/ModMuted.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/osu.Game/Rulesets/Mods/ModMuted.cs b/osu.Game/Rulesets/Mods/ModMuted.cs
--- a/osu.Game/Rulesets/Mods/ModMuted.cs	(revision 016e0c320745dae61a9c273d4bbb46d382392baa)
+++ b/osu.Game/Rulesets/Mods/ModMuted.cs	(date 1643959558362)
@@ -1,13 +1,15 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Linq;
 using osu.Framework.Audio;
-using osu.Framework.Audio.Track;
+using osu.Framework.Audio.Mixing;
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Sprites;
 using osu.Framework.Localisation;
+using osu.Game.Audio.Effects;
 using osu.Game.Configuration;
 using osu.Game.Graphics.UserInterface;
 using osu.Game.Overlays.Settings;
@@ -28,13 +30,14 @@
         public override double ScoreMultiplier => 1;
     }
 
-    public abstract class ModMuted<TObject> : ModMuted, IApplicableToDrawableRuleset<TObject>, IApplicableToTrack, IApplicableToScoreProcessor
+    public abstract class ModMuted<TObject> : ModMuted, IApplicableToDrawableRuleset<TObject>, IApplicableToScoreProcessor, IApplicableToTrackMixer
         where TObject : HitObject
     {
         private readonly BindableNumber<double> mainVolumeAdjust = new BindableDouble(0.5);
         private readonly BindableNumber<double> metronomeVolumeAdjust = new BindableDouble(0.5);
 
         private BindableNumber<int> currentCombo;
+        private AudioFilter lowPassFilter;
 
         [SettingSource("Enable metronome", "Add a metronome beat to help you keep track of the rhythm.")]
         public BindableBool EnableMetronome { get; } = new BindableBool
@@ -69,11 +72,11 @@
         protected ModMuted()
         {
             InverseMuting.BindValueChanged(i => MuteComboCount.MinValue = i.NewValue ? 1 : 0, true);
-        }
-
-        public void ApplyToTrack(ITrack track)
-        {
-            track.AddAdjustment(AdjustableProperty.Volume, mainVolumeAdjust);
+            mainVolumeAdjust.BindValueChanged(i =>
+            {
+                if (lowPassFilter != null)
+                    lowPassFilter.Cutoff = (int)(Math.Pow(i.NewValue, 3) * AudioFilter.MAX_LOWPASS_CUTOFF);
+            });
         }
 
         public void ApplyToDrawableRuleset(DrawableRuleset<TObject> drawableRuleset)
@@ -107,6 +110,11 @@
         }
 
         public ScoreRank AdjustRank(ScoreRank rank, double accuracy) => rank;
+
+        public void ApplyToTrackMixer(AudioMixer mixer)
+        {
+            lowPassFilter = new AudioFilter(mixer) { Cutoff = 1000 };
+        }
     }
 
     public class MuteComboSlider : OsuSliderBar<int>
```

</details>

hlysine committed:
Add IApplicableToTrackMixer

hlysine committed:
Fix naming for parameter

hlysine committed:
Update comment

hlysine committed:
Add interface for sample mixer

frenzibyte commented on osu.Game/Overlays/MusicController.cs:
```diff
@@ -389,11 +392,17 @@ public void ResetTrackAdjustments()
             CurrentTrack.RemoveAllAdjustments(AdjustableProperty.Frequency);
             CurrentTrack.RemoveAllAdjustments(AdjustableProperty.Tempo);
             CurrentTrack.RemoveAllAdjustments(AdjustableProperty.Volume);
+            audioManager.TrackMixer.Effects.Clear();
+            audioManager.SampleMixer.Effects.Clear();
```
> - **Should local mixers be provided to ApplyTo*Mixer instead?**
I don't think any local mixers exist yet. Just providing the global mixers should be the simplest working implementation IMO.

This is percisely the reason I stated to use a local mixer. Clearing all **global** mixer effects on `ResetTrackAdjustments` is definitely bad, as that may interrupt other effects added by other components.

You can add a `DrawableAudioMixer` layer on top of the `CurrentTrack` specific for mod effects, so you can clear the effects on `ResetTrackAdjustments` safely.

bdach replied:
I don't know that clearing all effects from the track mixer is all that bad in the context of this method at face value (note the pre-existing and indiscriminate removal of _all_ adjustments from the current track), but touching the samples mixer as well definitely seems dodgy as the method name does not mention samples.

frenzibyte replied:
In comparsion, the `RemoveAllAdjustments` call above affect the `CurrentTrack` component only, not the underlying `Beatmap.Value.Track` which certain components such as [`PlayerLoader`](https://github.com/ppy/osu/blob/98524d60a4a299ed10babd237515ca897c0b39fe/osu.Game/Screens/Play/PlayerLoader.cs#L208) and [`FailAnimation`](https://github.com/ppy/osu/blob/c6ae255326a07da5fc73ab088b4e4560715600f5/osu.Game/Screens/Play/FailAnimation.cs#L128-L129) apply adjustments to.

Meanwhile this clears all track/sample mixer effects applied on an `AudioManager` level, which is why I believe a local mixer is warranted here.

bdach replied:
Yeah as I said the test failures are 99% due to the effect clear so even if this is to be done as written the tests will need fixing first and foremost.

hlysine replied:
All the test failures originate from the assertion here:
https://github.com/ppy/osu/blob/fd81842adece837b198daf8bee81960003051aa4/osu.Game/Audio/Effects/AudioFilter.cs#L129-L137

I simply removed this assertion because removing effects from the audio mixer's side sounds like a perfectly valid use case for me, and in that situation, the audio filters will still think they are attached, failing the assertion.

Also, a `DrawableAudioMixer` has been added locally in the music controller. It is a simple addition so I can get behind using a local mixer here.

frenzibyte commented on osu.Game/Overlays/MusicController.cs:
```diff
@@ -389,11 +392,17 @@ public void ResetTrackAdjustments()
             {
                 foreach (var mod in mods.Value.OfType<IApplicableToTrack>())
                     mod.ApplyToTrack(CurrentTrack);
+                foreach (var mod in mods.Value.OfType<IApplicableToTrackMixer>())
+                    mod.ApplyToTrackMixer(audioManager.TrackMixer);
+                foreach (var mod in mods.Value.OfType<IApplicableToSampleMixer>())
+                    mod.ApplyToSampleMixer(audioManager.SampleMixer);
```
> - **Should IApplicableToTrack provide the mixer instead of a separate interface?**
@frenzibyte originally proposed to include the mixer in IApplicableToTrack.ApplyToTrack, but there are many cases where a mixer isn't available (e.g. when using this interface to calculate new track rate in diff calc), so I decided to split this into a new interface.

I still don't really see the reason behind these existing, you can add the mixer but as an optional parameter for such cases:
```csharp
public void ApplyToTrack(ITrack track, Mixer mixer = null);
```

And avoid all the extra additions of `IApplicableToTrackMixer`/`IApplicableToSampleMixer`/`IApplicableToMixers`.

bdach replied:
Why would the mixer be nullable in the above proposal? What would a mod implementation do if it is supplied a null mixer while it wants to apply effects, just give up?

I would even go a step further and propose that maybe `ITrack` should itself provide a reference to the mixer, rendering all of this unnecessary. But I don't feel confident on that proposal, so I'd advise not to take it too seriously.

frenzibyte replied:
> Why would the mixer be nullable in the above proposal? What would a mod implementation do if it is supplied a null mixer while it wants to apply effects, just give up?

It's nullable because it's irrelevant in cases like diffcalc. And, correct, I don't see that being a problem, we could not make it an optional parameter if your concern is about it being forgotten in future usages?

bdach replied:
Yeah I dunno. I think diffcalc doing stuff on tracks in general is kinda dodgy already, and it's currently worked around by passing a virtual track. I'd rather follow that lead and pass a dummy mixer or something, but the perfect scenario would be to decouple diffcalc from audio stuff.

frenzibyte replied:
Having a `VirtualAudioMixer` sounds workable, sure.

hlysine replied:
Created a framework PR for that: https://github.com/ppy/osu-framework/pull/5013

frenzibyte commented on osu.Game/Rulesets/Mods/IApplicableToSampleMixer.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace osu.Game.Rulesets.Mods
+{
+    /// <summary>
+    /// An interface for mods that add effects to the sample mixer.
+    /// </summary>
+    public interface IApplicableToSampleMixer : IApplicableMod
+    {
+        void ApplyToSampleMixer(AudioMixer mixer);
+    }
```
> - **Should IApplicableToSampleMixer exist?**
Yes for symmetry, but muted mod probably won't be needing this since it requires separate control of hit samples and metronome samples. Feel free to revert https://github.com/ppy/osu/commit/016e0c320745dae61a9c273d4bbb46d382392baa if this is deemed redundant.

I would say it's best not to provide a mixer for samples currently. I understand it's for symmetry, but it's not clear when a mod will actually want to apply an effect to the global sample mixer, or any specific sample mixers if at all.

hlysine replied:
Reverted the addition of `IApplicableToSampleMixer`

hlysine committed:
Revert "Add interface for sample mixer"

This reverts commit 016e0c320745dae61a9c273d4bbb46d382392baa.

hlysine committed:
Do not assert that mixer contains the filter in `ensureDetached`

All test failures originates from this assertion, which prevents mixers from clearing effects on their own. In practice, `AudioFilter` already gracefully handles such cases in `updateFilter`, so I don't think this assertion is necessary.

hlysine committed:
Expose `IAudioMixer` interface instead of `AudioMixer` abstract class

hlysine committed:
Use local audio mixer in `MusicController`

hlysine committed:
Use local mixer in `Player`

frenzibyte commented on osu.Game/Overlays/MusicController.cs:
```diff
@@ -52,14 +53,13 @@ public class MusicController : CompositeDrawable
 
         private readonly DrawableAudioMixer drawableAudioMixer;
 
+        public IAudioMixer AudioMixer => drawableAudioMixer;
```
nit: Would rather call this `TrackMixer` or just `Mixer` (same applies for the field, name feels too long). Also should have xmldoc pointing out its purpose (and that it's layered on top of `CurrentTrack`).

hlysine replied:
Done. Renamed it to `TrackMixer`.

hlysine committed:
Rename `AudioMixer` to `TrackMixer` and add xmldoc
